client/lib/streamframer/framer.go

```diff
@@ -177,40 +182,31 @@ OUTER:
 			}
 
 			// Read the data for the frame, and send it
-			s.f.Data = s.readData()
-			err = s.send(&s.f)
-			s.f.Clear()
+			s.send()
 			s.l.Unlock()
-			if err != nil {
-				return
-			}
 		case <-s.heartbeat.C:
 			// Send a heartbeat frame
-			if err = s.send(HeartbeatStreamFrame); err != nil {
-				return
+			select {
+			case s.out <- HeartbeatStreamFrame:
+			case <-s.exitCh:
 			}
 		}
 	}
 
 	s.l.Lock()
 	if !s.f.IsCleared() {
-		s.f.Data = s.readData()
-		err = s.send(&s.f)
-		s.f.Clear()
+		s.send()
 	}
 	s.l.Unlock()
 }
 
 // send takes a StreamFrame, encodes and sends it
-func (s *StreamFramer) send(f *StreamFrame) error {
-	sending := *f
-	f.Data = nil
```

Comment:
Don't let this fool you as it fooled me and everyone who touched it before! sweat_smile

Nil'ing the slice doesn't prevent a race. sending.Data will have the same backing array as f.Data used to. This means the next call to s.readData operates on the same backing array as sending.Data is reading in another goroutine!

At least that's my best explanation (and reverting this file immediately induces the bug).